File: ORA/app/src/main/java/com/otorunners/ora/auth/AuthValidation.kt
package com.otorunners.ora.auth

import java.util.Locale

fun pinValidationError(pin: String): String? {
    return when {
        pin.isEmpty() -> "Enter your PIN."
        pin.length != PIN_LENGTH -> "PIN must be exactly 4 digits."
        !pin.all(Char::isDigit) -> "PIN must contain numbers only."
        else -> null
    }
}

fun isValidPin(pin: String): Boolean = pinValidationError(pin) == null

fun normalizeNickname(nickname: String): String = nickname.trim()

fun canonicalNickname(nickname: String): String {
    return normalizeNickname(nickname).uppercase(Locale.ROOT)
}

fun nicknameValidationError(nickname: String): String? {
    val normalized = normalizeNickname(nickname)
    return when {
        normalized.isEmpty() -> "Enter a nickname."
        normalized.length > MAX_NICKNAME_LENGTH -> "Nickname can have up to 8 characters."
        !normalized.all(Char::isLetterOrDigit) -> "Use letters and numbers only."
        else -> null
    }
}

fun isValidNickname(nickname: String): Boolean = nicknameValidationError(nickname) == null

const val MAX_NICKNAME_LENGTH = 8
private const val PIN_LENGTH = 4

File: ora_flutter/lib/features/auth/domain/auth_validation.dart
const maxNicknameLength = 8;

String? nikValidationError(String nik) =>
    nik.trim().isEmpty ? 'Enter your NIK.' : null;

String? pinValidationError(String pin) {
  if (pin.isEmpty) return 'Enter your PIN.';
  if (pin.length != 4) return 'PIN must be exactly 4 digits.';
  if (!RegExp(r'^\d{4}$').hasMatch(pin)) {
    return 'PIN must contain numbers only.';
  }
  return null;
}

String normalizeNickname(String nickname) => nickname.trim();

String canonicalNickname(String nickname) =>
    normalizeNickname(nickname).toUpperCase();

String? nicknameValidationError(String nickname) {
  final normalized = normalizeNickname(nickname);
  if (normalized.isEmpty) return 'Enter a nickname.';
  if (normalized.length > maxNicknameLength) {
    return 'Nickname can have up to 8 characters.';
  }
  if (!RegExp(r'^[A-Za-z0-9]+$').hasMatch(normalized)) {
    return 'Use letters and numbers only.';
  }
  return null;
}
